// readmeio/dist/src/lib/get-project-base-url.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.getProjectBaseUrl = exports.getCache = void 0;
var crypto_1 = __importDefault(require("crypto"));
var find_cache_dir_1 = __importDefault(require("find-cache-dir"));
var flat_cache_1 = __importDefault(require("flat-cache"));
var node_fetch_1 = __importDefault(require("node-fetch"));
var timeout_signal_1 = __importDefault(require("timeout-signal"));
var package_json_1 = __importDefault(require("../../package.json"));
var config_1 = __importDefault(require("../config"));
function getCache(readmeApiKey) {
    var encodedApiKey = Buffer.from("".concat(readmeApiKey, ":")).toString('base64');
    var cacheDir = (0, find_cache_dir_1["default"])({ name: package_json_1["default"].name, create: true });
    var fsSafeApikey = crypto_1["default"].createHash('md5').update(encodedApiKey).digest('hex');
    // Since we might have differences of cache management, set the package version into the cache key so all caches will
    // automatically get refreshed when the package is updated/installed.
    var cacheKey = "".concat(package_json_1["default"].name, "-").concat(package_json_1["default"].version, "-").concat(fsSafeApikey);
    return flat_cache_1["default"].load(cacheKey, cacheDir);
}
exports.getCache = getCache;
function getProjectBaseUrl(readmeApiKey, requestTimeout) {
    if (requestTimeout === void 0) { requestTimeout = config_1["default"].timeout; }
    return __awaiter(this, void 0, void 0, function () {
        var encodedApiKey, cache, lastUpdated, signal_1, baseUrl_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    encodedApiKey = Buffer.from("".concat(readmeApiKey, ":")).toString('base64');
                    cache = getCache(readmeApiKey);
                    lastUpdated = cache.getKey('lastUpdated');
                    if (!(lastUpdated === undefined ||
                        (lastUpdated !== undefined && Math.abs(lastUpdated - Math.round(Date.now() / 1000)) >= 86400))) return [3 /*break*/, 2];
                    signal_1 = (0, timeout_signal_1["default"])(requestTimeout);
                    return [4 /*yield*/, (0, node_fetch_1["default"])("".concat(config_1["default"].readmeApiUrl, "/v1/"), {
                            method: 'get',
                            headers: {
                                Authorization: "Basic ".concat(encodedApiKey),
                                'User-Agent': "".concat(package_json_1["default"].name, "/").concat(package_json_1["default"].version)
                            },
                            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                            // @ts-ignore
                            signal: signal_1
                        })
                            .then(function (res) {
                            if (res.status >= 400 && res.status <= 599) {
                                throw res;
                            }
                            return res.json();
                        })
                            .then(function (project) {
                            baseUrl_1 = project.baseUrl;
                            cache.setKey('baseUrl', project.baseUrl);
                            cache.setKey('lastUpdated', Math.round(Date.now() / 1000));
                        })["catch"](function () {
                            // If unable to access the ReadMe API for whatever reason, let's set the last updated time to two minutes from
                            // now yesterday so that in 2 minutes we'll automatically make another attempt.
                            cache.setKey('baseUrl', null);
                            cache.setKey('lastUpdated', Math.round(Date.now() / 1000) - 86400 + 120);
                        })["finally"](function () {
                            timeout_signal_1["default"].clear(signal_1);
                        })];
                case 1:
                    _a.sent();
                    cache.save();
                    return [2 /*return*/, baseUrl_1];
                case 2: return [2 /*return*/, cache.getKey('baseUrl')];
            }
        });
    });
}
exports.getProjectBaseUrl = getProjectBaseUrl;
