// readmeio/dist/src/lib/log.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.log = void 0;
var url = __importStar(require("url"));
var clamp_1 = __importDefault(require("lodash/clamp"));
var uuid_1 = require("uuid");
var config_1 = __importDefault(require("../config"));
var construct_payload_1 = require("./construct-payload");
var get_project_base_url_1 = require("./get-project-base-url");
var is_request_1 = __importDefault(require("./is-request"));
var metrics_log_1 = require("./metrics-log");
var patch_request_1 = require("./patch-request");
var patch_response_1 = require("./patch-response");
var queue = [];
function doSend(readmeApiKey, options) {
    // Copy the queue so we can send all the requests in one batch
    var json = __spreadArray([], queue, true);
    // Clear out the queue so we don't resend any data in the future
    queue = [];
    // Make the log call
    (0, metrics_log_1.metricsAPICall)(readmeApiKey, json)["catch"](function (e) {
        // Silently discard errors and timeouts.
        if (options.development)
            throw e;
    });
}
// Make sure we flush the queue if the process is exited
process.on('exit', doSend);
function setDocumentationHeader(res, baseLogUrl, logId) {
    // This is to catch the potential race condition where `getProjectBaseUrl()`
    // takes longer to respond than the original req/res to finish. Without this
    // we would get an error that would be very difficult to trace. This could
    // do with a test, but it's a little difficult to test. Maybe with a nock()
    // delay timeout.
    if (res.headersSent)
        return;
    res.setHeader('x-documentation-url', "".concat(baseLogUrl, "/logs/").concat(logId));
}
/**
 * This method will send supplied API requests to ReadMe Metrics.
 *
 * @see {@link https://readme.com/metrics}
 * @see {@link https://docs.readme.com/docs/sending-logs-to-readme-with-nodejs}
 * @param readmeApiKey The API key for your ReadMe project. This ensures your requests end up in
 *    your dashboard. You can read more about the API key in
 *    [our docs](https://docs.readme.com/reference/authentication).
 * @param req This is your incoming request object from your HTTP server and/or framework.
 * @param res This is your outgoing response object for your HTTP server and/or framework.
 * @param group A function that helps translate incoming request data to our metrics grouping data.
 * @param options Additional options. See the documentation for more details.
 */
function log(readmeApiKey, req, res, group, options) {
    if (options === void 0) { options = {}; }
    if (!readmeApiKey)
        throw new Error('You must provide your ReadMe API key');
    if (!group)
        throw new Error('You must provide a group');
    // Ensures the buffer length is between 1 and 30
    var bufferLength = (0, clamp_1["default"])(options.bufferLength || config_1["default"].bufferLength, 1, 30);
    var startedDateTime = new Date();
    var logId = (0, uuid_1.v4)();
    // baseLogUrl can be provided, but if it isn't then we
    // attempt to fetch it from the ReadMe API
    if (typeof options.baseLogUrl === 'string') {
        setDocumentationHeader(res, options.baseLogUrl, logId);
    }
    else {
        (0, get_project_base_url_1.getProjectBaseUrl)(readmeApiKey).then(function (baseLogUrl) {
            setDocumentationHeader(res, baseLogUrl, logId);
        });
    }
    (0, patch_response_1.patchResponse)(res);
    (0, patch_request_1.patchRequest)(req);
    /*
     * This should in future become more sophisticated, with flush timeouts and more error checking but
     * this is fine for now
     */
    function startSend() {
        var requestBody = req.body;
        if ((0, is_request_1["default"])(req, 'text/*')) {
            if ('_text' in req) {
                requestBody = req._text;
            }
        }
        else if ((0, is_request_1["default"])(req, '+json')) {
            if ('_json' in req) {
                requestBody = req._json;
            }
        }
        else if ((0, is_request_1["default"])(req, 'application/x-www-form-urlencoded')) {
            if ('_form_encoded' in req) {
                requestBody = req._form_encoded;
            }
        }
        var payload = (0, construct_payload_1.constructPayload)(req, res, __assign(__assign({}, group), { logId: logId, startedDateTime: startedDateTime, responseEndDateTime: new Date(), routePath: req.route
                ? url.format({
                    protocol: req.protocol,
                    host: req.hostname,
                    pathname: "".concat(req.baseUrl).concat(req.route.path)
                })
                : '', responseBody: res._body, requestBody: requestBody }), options);
        queue.push(payload);
        if (queue.length >= bufferLength)
            doSend(readmeApiKey, options);
        cleanup(); // eslint-disable-line @typescript-eslint/no-use-before-define
    }
    function cleanup() {
        res.removeListener('finish', startSend);
        res.removeListener('error', cleanup);
        res.removeListener('close', cleanup);
    }
    // Add response listeners
    res.once('finish', startSend);
    res.once('error', cleanup);
    res.once('close', cleanup);
    return logId;
}
exports.log = log;
