// readmeio/dist/src/lib/metrics-log.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.metricsAPICall = exports.setBackoff = void 0;
var node_fetch_1 = __importDefault(require("node-fetch"));
var timeout_signal_1 = __importDefault(require("timeout-signal"));
var package_json_1 = __importDefault(require("../../package.json"));
var config_1 = __importDefault(require("../config"));
var BACKOFF_SECONDS = 15; // when we need to backoff HTTP requests, pause for seconds
var backoffExpiresAt;
// Exported for use in unit tests
function setBackoff(expiresAt) {
    backoffExpiresAt = expiresAt;
}
exports.setBackoff = setBackoff;
function shouldBackoff(response) {
    // Some HTTP error codes indicate a problem with the API key, like the key is
    // invalid or it's being rate limited. To avoid pointless requests to the
    // ReadMe server, pause outgoing requests for a few seconds before trying
    // again. Logs will be silently discarded while requests are paused, which is
    // acceptable since the server wouldn't accept them anyway.
    switch (response.status) {
        case 401: // Unauthorized, i.e. this API key is invalid
            return true;
        case 403: // Forbidden, i.e. this API key is blocked by the server
            return true;
        case 429: // Too Many Requests, i.e. this API key is currently being rate limited
            return true;
        case 500: // Internal Server Error; give the ReadMe server a chance to recover
            return true;
        case 503: // Service Unavailable; same thing
            return true;
        default:
            return false;
    }
}
function getLogIds(body) {
    if (Array.isArray(body)) {
        return body.map(function (value) { return value._id; });
    }
    return body._id;
}
function metricsAPICall(readmeAPIKey, body, fireAndForget) {
    if (fireAndForget === void 0) { fireAndForget = false; }
    var signal = (0, timeout_signal_1["default"])(config_1["default"].timeout);
    var encodedKey = Buffer.from("".concat(readmeAPIKey, ":")).toString('base64');
    var makeRequest = function () {
        if (backoffExpiresAt) {
            if (backoffExpiresAt > new Date()) {
                return Promise.resolve();
            }
            // after the backoff expires, erase the old expiration time
            backoffExpiresAt = undefined;
        }
        return (0, node_fetch_1["default"])(new URL('/v1/request', config_1["default"].host), {
            method: 'post',
            body: JSON.stringify(body),
            headers: {
                Authorization: "Basic ".concat(encodedKey),
                'Content-Type': 'application/json',
                'User-Agent': "".concat(package_json_1["default"].name, "/").concat(package_json_1["default"].version)
            },
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            signal: signal
        })
            .then(function (response) {
            if (shouldBackoff(response)) {
                // backoff for a few seconds, but not if another callback has already started backing off
                if (!backoffExpiresAt) {
                    backoffExpiresAt = new Date();
                    backoffExpiresAt.setSeconds(backoffExpiresAt.getSeconds() + BACKOFF_SECONDS);
                }
            }
            return response;
        })["finally"](function () {
            timeout_signal_1["default"].clear(signal);
        });
    };
    if (fireAndForget) {
        makeRequest();
        return Promise.resolve({
            ids: getLogIds(body)
        });
    }
    return makeRequest().then(function (response) {
        return {
            response: response,
            ids: getLogIds(body)
        };
    });
}
exports.metricsAPICall = metricsAPICall;
