// readmeio/dist/src/lib/process-response.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var http_1 = require("http");
var omit_1 = __importDefault(require("lodash/omit"));
var pick_1 = __importDefault(require("lodash/pick"));
var object_to_array_1 = require("./object-to-array");
var process_request_1 = require("./process-request");
/**
 * Transforms the provided ServerResponse and additional information into the appropriate HAR structure
 *
 * @param res The node ServerResponse object
 * @param responseBody A string representation of the response body
 * @param options A collection of additional options. See the documentation for more details.
 *
 * @returns The HAR formatted response details
 */
function processResponse(res, responseBody, options) {
    var denylist = (options === null || options === void 0 ? void 0 : options.denylist) || (options === null || options === void 0 ? void 0 : options.blacklist);
    var allowlist = (options === null || options === void 0 ? void 0 : options.allowlist) || (options === null || options === void 0 ? void 0 : options.whitelist);
    var body;
    try {
        body = JSON.parse(responseBody);
        // Only apply blacklist/whitelist if it's an object
        if (denylist) {
            body = (0, omit_1["default"])(body, denylist);
        }
        if (allowlist && !denylist) {
            body = (0, pick_1["default"])(body, allowlist);
        }
    }
    catch (e) {
        // Non JSON body
        body = responseBody;
    }
    var headers = res.getHeaders();
    if (denylist) {
        headers = (0, omit_1["default"])(headers, denylist);
    }
    if (allowlist && !denylist) {
        headers = (0, pick_1["default"])(headers, allowlist);
    }
    return {
        status: res.statusCode,
        // In fastify, at the point where we have to fetch the statusMessage
        // from the response, it is not set because the headers haven't been
        // flushed yet, so we need to fetch the default value from Node.
        // https://nodejs.org/dist/latest-v14.x/docs/api/http.html#http_response_statusmessage
        //
        // This is the same thing that Node.js does internally:
        // https://github.com/nodejs/node/blob/9b8ba2536044ae08a1cd747a3aa52df7d1815e7e/lib/_http_server.js#L318
        statusText: res.statusMessage || http_1.STATUS_CODES[res.statusCode],
        headers: (0, object_to_array_1.objectToArray)(headers, { castToString: true }),
        content: {
            text: JSON.stringify(body),
            size: Number((0, process_request_1.fixHeader)(res.getHeader('content-length') || 0)),
            mimeType: (0, process_request_1.fixHeader)(res.getHeader('content-type')) || 'text/plain'
        },
        // TODO: Once readme starts accepting these, send the correct values
        httpVersion: '',
        cookies: [],
        redirectURL: '',
        headersSize: 0,
        bodySize: 0
    };
}
exports["default"] = processResponse;
