// readmeio/dist/src/lib/process-request.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.fixHeader = void 0;
var qs = __importStar(require("querystring"));
var url_1 = __importStar(require("url"));
var contentType = __importStar(require("content-type"));
var get_1 = __importDefault(require("lodash/get"));
var merge_1 = __importDefault(require("lodash/merge"));
var pick_1 = __importDefault(require("lodash/pick"));
var set_1 = __importDefault(require("lodash/set"));
var construct_payload_1 = require("./construct-payload");
var object_to_array_1 = require("./object-to-array");
/**
 * Ensure we have a string or undefined response for any header.
 *
 * @param header
 * @returns
 */
function fixHeader(header) {
    if (header === undefined) {
        return undefined;
    }
    if (Array.isArray(header)) {
        return header.join(',');
    }
    return String(header);
}
exports.fixHeader = fixHeader;
/**
 * Redacts a value by replacing it with a string like [REDACTED 6]
 *
 * @param value the value to be redacted
 * @returns A redacted string potentially containing the length of the original value, if it was a string
 */
function redactValue(value) {
    var redactedVal = typeof value === 'string' ? " ".concat(value.length) : '';
    return "[REDACTED".concat(redactedVal, "]");
}
/**
 * Redacts all the properties in an object
 *
 * @param obj The data object that is operated upon
 * @param redactedPaths a list of paths that point values which should be redacted
 * @returns An object with the redacted values
 */
function redactProperties(obj, redactedPaths) {
    if (redactedPaths === void 0) { redactedPaths = []; }
    var nextObj = __assign({}, obj);
    return redactedPaths.reduce(function (acc, path) {
        var value = (0, get_1["default"])(acc, path);
        if (value !== undefined)
            (0, set_1["default"])(acc, path, redactValue(value));
        return acc;
    }, nextObj);
}
/**
 * @param obj The data object that is operated upon
 * @param cb A callback that is invoked for each value found, the return value being the next value that is set in the returned object
 * @returns An object with the replaced values
 */
function replaceEach(obj, cb) {
    return Object.keys(obj).reduce(function (acc, key) {
        var value = obj[key];
        if (typeof value === 'object' && value !== null) {
            acc[key] = replaceEach(value, cb);
        }
        else if (value !== undefined) {
            acc[key] = cb(value);
        }
        return acc;
    }, {});
}
/**
 * Redacts everything but the provided fields
 *
 * @param obj The data object with fields to redact
 * @param nonRedactedPaths A list of all object paths that shouldn't be redacted
 * @returns A merged objects that is entirely redacted except for the values of the nonRedactedPaths
 */
function redactOtherProperties(obj, nonRedactedPaths) {
    var allowedFields = (0, pick_1["default"])(obj, nonRedactedPaths);
    var redactedFields = obj ? replaceEach(obj, redactValue) : obj;
    return (0, merge_1["default"])(redactedFields, allowedFields);
}
function isApplicationJson(mimeType) {
    if (!mimeType) {
        return false;
    }
    return (['application/json', 'application/x-json', 'text/json', 'text/x-json'].includes(mimeType) ||
        mimeType.indexOf('+json') !== -1);
}
function parseRequestBody(body, mimeType) {
    if (mimeType === 'application/x-www-form-urlencoded') {
        return qs.parse(body);
    }
    if (isApplicationJson(mimeType)) {
        try {
            return JSON.parse(body);
        }
        catch (err) {
            // no-op
        }
    }
    return body;
}
/**
 * This transforms the IncommingMessage and additional provided information into the relevant HAR structure
 *
 * @param req The IncommingMessage from the node server.
 * @param requestBody A parsed request body object, or an unparsed request body string.
 * @param options A collection of processing options.
 *
 * @returns The proper request structure following the HAR format
 */
function processRequest(req, requestBody, options) {
    var _a;
    var protocol = ((_a = fixHeader(req.headers['x-forwarded-proto'])) === null || _a === void 0 ? void 0 : _a.toLowerCase()) || (0, construct_payload_1.getProto)(req);
    var host = fixHeader(req.headers['x-forwarded-host']) || req.headers.host;
    var denylist = (options === null || options === void 0 ? void 0 : options.denylist) || (options === null || options === void 0 ? void 0 : options.blacklist);
    var allowlist = (options === null || options === void 0 ? void 0 : options.allowlist) || (options === null || options === void 0 ? void 0 : options.whitelist);
    var mimeType = null;
    try {
        mimeType = contentType.parse(req.headers['content-type']).type;
    }
    catch (e) { } // eslint-disable-line no-empty
    var reqBody = typeof requestBody === 'string' ? parseRequestBody(requestBody, mimeType) : requestBody;
    var postData = null;
    if (denylist) {
        reqBody = typeof reqBody === 'object' ? redactProperties(reqBody, denylist) : reqBody;
        req.headers = redactProperties(req.headers, denylist);
    }
    if (allowlist && !denylist) {
        reqBody = typeof reqBody === 'object' ? redactOtherProperties(reqBody, allowlist) : reqBody;
        req.headers = redactOtherProperties(req.headers, allowlist);
    }
    if (mimeType === 'application/x-www-form-urlencoded') {
        postData = {
            mimeType: mimeType,
            // `reqBody` is likely to be an object, but can be empty if no HTTP body sent
            params: (0, object_to_array_1.objectToArray)((reqBody || {}))
        };
    }
    else if (isApplicationJson(mimeType)) {
        postData = {
            mimeType: mimeType,
            text: typeof reqBody === 'object' || Array.isArray(reqBody) ? JSON.stringify(reqBody) : reqBody
        };
    }
    else if (mimeType) {
        var stringBody = '';
        try {
            stringBody = typeof reqBody === 'string' ? reqBody : JSON.stringify(reqBody);
        }
        catch (e) {
            stringBody = '[ReadMe is unable to handle circular JSON. Please contact support if you have any questions.]';
        }
        postData = {
            mimeType: mimeType,
            // Do our best to record *some sort of body* even if it's not 100% accurate.
            text: stringBody
        };
    }
    // We use a fake host here because we rely on the host header which could be redacted.
    // We only ever use this reqUrl with the fake hostname for the pathname and querystring.
    // req.originalUrl is express specific, req.url is node.js
    var reqUrl = new url_1.URL(req.originalUrl || req.url, 'https://readme.io');
    if (req.headers.authorization) {
        req.headers.authorization = (0, construct_payload_1.mask)(req.headers.authorization);
    }
    var requestData = {
        method: req.method,
        url: url_1["default"].format({
            // Handle cases where some reverse proxies put two protocols into x-forwarded-proto
            // This line does the following: "https,http" -> "https"
            // https://github.com/readmeio/metrics-sdks/issues/378
            protocol: protocol.split(',')[0],
            host: host,
            pathname: reqUrl.pathname,
            // Search includes the leading questionmark, format assumes there isn't one, so we trim that off.
            query: qs.parse(reqUrl.search.substring(1))
        }),
        httpVersion: "".concat((0, construct_payload_1.getProto)(req).toUpperCase(), "/").concat(req.httpVersion),
        headers: (0, object_to_array_1.objectToArray)(req.headers),
        queryString: (0, object_to_array_1.searchToArray)(reqUrl.searchParams),
        postData: postData,
        // TODO: When readme starts accepting these, send the correct values
        cookies: [],
        headersSize: -1,
        bodySize: -1
    };
    if (requestData.postData === null) {
        delete requestData.postData;
    }
    return requestData;
}
exports["default"] = processRequest;
